src/memory_extension.rs

```diff
@@ -165,43 +191,165 @@ impl MemoryExtension for CacheLru {
+                    let (_, oldline_content) = self.cache.pop_at(0).unwrap();
+                    oldline_content
+                } else {
+                    // 2b. Add new line
+                    [0; MAX_BUS_SIZE]
+                }
+            };
+            scadata.cache_updates.push((oldline_content, memory_before));
             self.cache.push((address, memory_before));
 
-            // Update cache line and bus if necessary
-            if memory_before != memory_after {
-                let last = self.cache.last_mut().unwrap();
-                scadata.cache_updates.push((last.1, memory_after));
-                last.1 = memory_after;
-                scadata.bus_updates.push((self.bus, memory_after));
-                self.bus = memory_after;
-            }
+            // 3. Update cache line and bus if necessary. Leaks: Cache (may), Bus (may), Memory (may)
+            self.update_and_leak_cache_bus_memory(scadata, self.cache.len() - 1, memory_after);
         }
         trace!("Cache: {:x?}", self.cache);
+    }
+}
 
-        // Update memory
-        if memory_before != memory_after {
-            scadata.memory_updates.push((memory_before, memory_after));
+#[derive(Debug)]
+struct CacheLruWriteBackCacheLine {
+    address: u64,
+    content: [u8; MAX_BUS_SIZE],
+    memory: [u8; MAX_BUS_SIZE],
+}
+
+pub struct CacheLruWriteBack {
+    bus_size: usize,
+    bus: [u8; MAX_BUS_SIZE],
+    cache_lines: usize,
+    cache: ArrayVec<CacheLruWriteBackCacheLine, MAX_CACHE_LINES>,
+}
+
+impl CacheLruWriteBack {
+    pub fn new(bus_size: usize, cache_lines: usize) -> Self {
+        Self {
+            bus_size,
+            bus: [0; MAX_BUS_SIZE],
+            cache_lines,
+            cache: ArrayVec::new(),
+        }
+    }
+
+    /// Update bus. Leaks: Bus
+    fn update_and_leak_bus(&mut self, scadata: &mut ScaData, newvalue: [u8; MAX_BUS_SIZE]) {
+        scadata.bus_updates.push((self.bus, newvalue));
+        self.bus = newvalue;
+    }
+
+    /// Update cache line with new content.
+    /// Leaks: Cache (may)
+    fn update_and_leak_cache(
+        &mut self,
+        scadata: &mut ScaData,
+        index: usize,
+        newcontent: [u8; MAX_BUS_SIZE],
+    ) {
+        let line = &self.cache[index];
+        if line.content == newcontent {
+            return;
+        }
+
+        scadata.cache_updates.push((line.content, newcontent));
+        let newline = CacheLruWriteBackCacheLine {
+            address: line.address,
+            content: newcontent,
+            memory: line.memory,
+        };
+        self.cache.remove(index);
+        self.cache.push(newline);
+    }
+
+    /// Execute write-back:
+    /// 1. Pop lru cache line
+    /// 2. If line is dirty, leak bus and memory
+    fn write_back_and_leak(&mut self, scadata: &mut ScaData) -> CacheLruWriteBackCacheLine {
+        let oldline = self.cache.pop_at(0).unwrap();
+        if oldline.content != oldline.memory {
```

Comment:
Good Point!
Storing the old memory is required so far because I didn't find a method to suppress the actual memory update done by unicorn.
E.g. when executing `str r0, [address]` unicorn will change the memory behind address regardless what I'm implementing in the cache.